formats/ctf-text/durin.c

```diff
@@ -0,0 +1,334 @@
+				&_cu_offset, NULL);
+	if (ret != DW_DLV_OK) {
+		goto error;
+	}
+	*cu_offset = _cu_offset;
+
+	return 0;
+
+error:
+	return -1;
+}
+
+/**
+ * Reset the internal CU offset state of `dwarf_info`, allowing to
+ * read from the first CU again.
+ *
+ * libdwarf keeps the value of the last read CU header's offset stored
+ * internally within the opaque Dwarf_Debug structure, from which it
+ * then computes the offset of the next header and fetches
+ * it. Therefore, the only apparent way to reset this offset is to
+ * iterate over all CUs until they wrap over, short of instantiating a
+ * new Dwarf_Debug structure. This function hides this manipulation.
+ *
+ * @param dwarf_info	Dwarf_Debug instance
+ * @returns		0 on success, -1 on failure.
+ */
+static
+int durin_cu_reset_internal_state(Dwarf_Debug *dwarf_info)
+{
+	int ret;
+	Dwarf_Unsigned cu_offset;
+
+	if (!dwarf_info) {
+		goto error;
+	}
+
+	do {
+		ret = dwarf_next_cu_header(*dwarf_info, NULL, NULL, NULL, NULL,
+					&cu_offset, NULL);
+	} while (ret == DW_DLV_OK);
+
+	if (ret == DW_DLV_ERROR) {
+		goto error;
+	}
+
+	return 0;
+
+error:
+	return -1;
+}
+
+struct durin_cu *durin_cu_create(Dwarf_Debug *dwarf_info)
+{
+	int ret;
+	struct durin_cu *cu;
+
+	if (!dwarf_info) {
+		goto error;
+	}
+
+	ret = durin_cu_reset_internal_state(dwarf_info);
+	if (ret) {
+		goto error;
+	}
+
+	cu = g_new0(struct durin_cu, 1);
+	if (!cu) {
+		goto error;
+	}
+	cu->dwarf_info = dwarf_info;
+	cu->offset = 0;
+
+	return cu;
+
+error:
+	return NULL;
+}
+
+void durin_cu_destroy(struct durin_cu *cu)
+{
+	g_free(cu);
+}
+
+int durin_cu_next(struct durin_cu *cu)
+{
+	int ret;
+	Dwarf_Unsigned cu_offset;
+
+	if (!cu) {
+		goto error;
+	}
+
+	ret = durin_cu_next_header(cu->dwarf_info, &cu_offset);
+	if (ret) {
+		goto error;
+	}
+	cu->offset = cu_offset;
+
+	return 0;
+
+error:
+	return -1;
+}
+
+struct durin_die *durin_die_create(struct durin_cu *cu)
+{
+	int ret;
+	struct durin_die *die = NULL;
+	Dwarf_Die *dwarf_die = NULL;
+	Dwarf_Error error;
+
+	if (!cu) {
+		goto error;
+	}
+
+	dwarf_die = g_new0(Dwarf_Die, 1);
+	if (!dwarf_die) {
+		goto error;
+	}
+	/* dwarf_siblingof on the cu fetches its root DIE. */
+	ret = dwarf_siblingof(*cu->dwarf_info, NULL, dwarf_die, &error);
+	if (ret != DW_DLV_OK) {
+		goto error;
+	}
+	die = g_new0(struct durin_die, 1);
+	if (!die) {
+		goto error;
+	}
+	die->cu = cu;
+	die->dwarf_die = dwarf_die;
+	die->depth = 0;
+
+	return die;
+
+error:
+	if (dwarf_die) {
+		dwarf_dealloc(*cu->dwarf_info, *dwarf_die, DW_DLA_DIE);
+		g_free(dwarf_die);
+	}
+	g_free(die);
+	return NULL;
+}
+
+void durin_die_destroy(struct durin_die *die)
+{
+	if (!die) {
+		return;
+	}
+
+	if (die->dwarf_die) {
+		dwarf_dealloc(*die->cu->dwarf_info, *die->dwarf_die,
+			      DW_DLA_DIE);
+		g_free(die->dwarf_die);
+	}
+	g_free(die);
+}
+
+int durin_die_next(struct durin_die *die)
+{
+	int ret;
+	Dwarf_Die *next_die = NULL;
+	Dwarf_Error error;
```

Comment:
I'm surprised this didn't clash with the "error" label since GCC allows the use of lables as values...

https://gcc.gnu.org/onlinedocs/gcc/Labels-as-Values.html